openfisca_core/formulas.py

```diff
@@ -433,162 +481,61 @@ def compute(self, period, **parameters):
                     simulation.stringify_input_variables_infos(input_variables_infos), str(period),
                     stringify_array(array)))
 
-        dated_holder = holder.put_in_cache(array, period, extra_params)
-
         self.clean_cycle_detection_data()
         if max_nb_cycles is not None:
             simulation.max_nb_cycles = None
 
-        return dated_holder
+        return holders.DatedHolder(self.holder, period, array, extra_params)
+
+    def find_function(self, period):
+        """
+        This function finds the last active formula for the time interval [period starting date, variable end attribute].
+        """
+        end = self.holder.column.end
+        if end and period.start.date > get_datetime_date(self, end):
+            return None
+
+        # Assume that self.dated_formulas and self.dated_formulas_class
+        # contain same formulas (instance and class) in same order
+        assert len(self.dated_formulas) == len(self.dated_formulas_class)
```

Comment:
If later a developer changes only `self.dated_formulas` or only`self.dated_formulas_class`, an error should be thrown to express that this important assumption is broken. Besides, it's the minimal assertion.